### ISSUE TITLE
Notify web-handler about client disconnection

### ISSUE THREAD

**Issue by asvetlov** [aio-libs/aiohttp#447]
I've received email from David Markey with ask for the subject.

> this is a HTTP long poll/push design (I know websockets is the way to go but this is a corner case)
> 
> When the client connects, a Queue and Redis PUB/SUB top is created to listen for data. Of course if the client abruptly disconnects these resources need to be cleaned up.

From my understanding the problem is: we have `start()` coroutine which process server connection. On disconnection transport is closed by parser code but it doesn't cancel the coroutine explicitly. It may be informed about cancellation only on stream reading. Please correct me if I'm wrong.

@fafhrd91 what do you think about?

**Comment by dmarkey** [comment]
With tornado I had the on_close method that I could override.

**Comment by fafhrd91** [comment]
Signal maybe? As I understand we just need to notify client code and not alter close behavior?

**Comment by asvetlov** [comment]
I think it should be task cancellation. Using try/finally for resource releasing looks more native than callback registering.

**Comment by fafhrd91** [comment]
I think I misunderstand the problem, could you provide how it could look like?

**Comment by dmarkey** [comment]
``` python
@asyncio.coroutine
def lphandler(request):
    text = yield from request.text()
    topic = queue = res = None
    try:
        topic, queue = yield from init_request(text, request)
        res = yield from wait_for_data(topic, queue)
    except ClientDisconnectedError:
        yield from close_connection(topic, queue)

    return res
```

**Comment by dmarkey** [comment]
That how I envisage it working kind of.

**Comment by asvetlov** [comment]
Actually it will be `CancelledError`: task has `.cancel()` method but it's impossible to send custom exception into waiting task.
@dmarkey I believe for you doesn't matter what exception to use for resource cleanup. Moreover usually you want just `try/finally`, not `try/except`.

**Comment by fafhrd91** [comment]
-10

you guys want to break try/except semantics.
from example it is natural to think that `res = yield from wait_for_data(topic, queue)` raises ClientDisconnectError.

@asvetlov do you remember asyncio discussion around proper task cancellation? initially asyncio has similar semantics for connection, it was not clear who cancel task. and that is super bad behavior.

**Comment by asvetlov** [comment]
@fafhrd91 please point on asyncio discussion you are mentioned. A link will be very useful.
from example it is natural to think that `res = yield from wait_for_data(topic, queue)` raises ClientDisconnectError. 
I remember something but not convinced that we have the same use case.

Adding callback (signal is a kind of callback, sure) for resource cleanup is a solution, but I'd like to keep web-handler code linear. `try/finally` is still native for the task.

Long story short: I'm ok with signal but want to make sure that there is no better way.

**Comment by fafhrd91** [comment]
it is looks natural, but it is very unnatural, you get exception from outside and not from inside of `try/except` 

i can not point to specific discussion, but i remember that because we met with Guido specifically to discuss that problem.

i think i found it, https://groups.google.com/forum/#!searchin/python-tulip/nikolay/python-tulip/EgpBV5-sIQ4/HdNgaVwBTPsJ

**Comment by dmarkey** [comment]
Is there anything I can use as a workaround? A socket disconnect callback or something?

**Comment by kxepal** [comment]
@dmarkey quite trivial: create channel/queue or whatever you need on connect, generate keep-alives events while client is active, run cleanup routines when ttl expires. This solution is even better than proposed in topic start since it allows to keep clients context it they suffers short-time disconnects (hi, mobile networks) or regular dynamic IP address update.

**Comment by filmor** [comment]
I'm looking for something like this, too. I'm afraid I didn't understand @kxepal's suggestion. My use-case is that I submit a calculation to a `ProcessPool` and wait for a result from the future object. If the connection is closed during that I want to `cancel` the future (basically raising `KeyboardInterruptError` in the worker process). In Tornado I have also implemented this using `on_close`, storing the future in a variable that is accessible to the application.

How would I implement this in aiohttp? So far I can only see overriding the `RequestHandlerFactory` and messing around in there ...

**Comment by asvetlov** [comment]
Well, we may add `on_disconnect` signal to `web.Request`.
`web-handler` may subscribe to the signal if it want to receive disconnection notification.

Another option is just adding `web.Request.disconnected` flag: the most widespread code does something like this:

```
async def handler(req):
   resp = web.Response()
   await resp.prepare(req)
   while True:
       data = await read_data_from_other_source()
       if not data:
           return resp
       if not req.disconnected:
           resp.write(data)
```

We can support both `disconnected` flag and `on_disconnect` signal -- the later may be useful to stop reading from outer source early.

@fafhrd91 thoughts?

**Comment by fafhrd91** [comment]
I am closing this issue, no need for signal or any new machinery.

`asyncio.CancelledError` means client disconnection.

**Comment by asvetlov** [comment]
Ok

**Comment by filmor** [comment]
@fafhrd91 Could you give me an example how I could use this? If I only receive this error on send I've probably already wasted a lot of time.

**Comment by fafhrd91** [comment]
```
def handler(req):
    try:
        resp = yield from actual_handler(req)
    except asyncio.CancelledError:
        # client disconnected
        raise

```
